### DISCUSSION TITLE
time based conditions

### DISCUSSION THREAD

**Discussion by bafnarahul2** [kernc/backtesting.py#761] [General]
How do i put time based conditions such as no trade after a particular period or all the positions to be closed 10 minutes before call.

**Comment by kernc** [comment]
Examine [`Strategy.data.index[-1]`](https://kernc.github.io/backtesting.py/doc/backtesting/backtesting.html#backtesting.backtesting.Strategy.data) which would be a [`pd.Timestamp`](https://pandas.pydata.org/docs/reference/api/pandas.Timestamp.html) for datetime indexes.

**Comment by acierum** [comment]
I'm trying to implement a condition where a trade is entered only between 8-10am, but the trade stats still show an entry at 14:00:00.  Not sure what I'm doing wrong here? Any input much appreciated!

```
class MACDStrategy(Strategy):
    def init(self):
        self.macd_hist = self.I(MACD, self.data.df) #indicator function to compute first
        self.atr = self.I(ATR, self.data.df)

    def next(self):
        high = self.data.High[-1]
        low = self.data.Low[-1]

        # buy the asset
        hour = self.data.index.hour
        if (6 < self.macd_hist[-1] > self.macd_hist[-2] and
            8 <= hour[-1] <= 10):
                if not self.position:
                    sl = low - self.atr * 1.5 # tp = price + self.atr * 2
                    self.buy(limit=high, stop=high, sl = sl) #buy if high exceeds prior high
        # Else sell/close the asset
        for trade in self.trades:
            if self.macd_hist[-1] < self.macd_hist[-2]:
                if self.position:
                    self.position.close()

#get historical data
df = pd.read_csv('MESH31HJan.csv', index_col='Date', parse_dates=True)

#Backtesting
bt = Backtest(df, MACDStrategy, cash=5_000, commission=0, exclusive_orders=True)
stats = bt.run()
```

<img width="900" alt="Screen Shot 2023-02-17 at 10 37 28 AM" src="https://user-images.githubusercontent.com/124842013/219698377-e205612a-fee3-4421-ab29-6848be4765f4.png">

**Reply by kernc** [reply]
You're buying with stop-limit. Can you confirm 14:00 bar is not the soonest the stop is reached?

**Reply by acierum** [reply]
Yes, you're absolutely correct! 

How can I avoid this situation if I just want to place trades within the specified hours?

**Reply by kernc** [reply]
Placed orders are [GTC](https://www.investopedia.com/terms/g/gtc.asp). Therefore you just have to make sure you cancel the order when you no longer want it filled. You can e.g. close the orders if they haven't been filled by the next bar's completion:
```py
def next(self):
   for order in self.orders:
       if not order.is_contingent:
           order.cancel()
   ...  # Rest of code
```

**Comment by RaccoonTrading** [comment]
Hi, posting here because I have a similar issue:

Coded the following custom indicator: It return 1 if within the timeframe, 0 if not within:
(i have tried with changing the astype types --> no bueno

```
`    def window(data):
        window = pd.Series(
            (data.index.time >= pd.to_datetime('14:29').time()) & (data.index.time <= pd.to_datetime('16:30').time()),
            index=data.index).astype(int)
        return np.array(window)`
```

It prints on the graph everything seems ok there. Values are correct. Nonetheless, it seems from time to time that the algo gets a trade before or after. and I will burst a nut if I don't find out why:

`            self.window = self.I(window, self.data, overlay=False)`

Here 9pm purchase UTC, with a window value to 0

<img width="731" alt="Screenshot 2024-01-20 at 19 56 52" src="https://github.com/kernc/backtesting.py/assets/155924232/9b650c52-c9ad-4a7d-a371-4355a7a28d5a">




I use it to close positon - which seems to work BTW, cause when it opens outside of the window, it close the position at the next bar:

```
            if self.position:
                if self.position.is_long:
                    if (
                            self.position.pl_pct <= max_day_loss
                            or self.data.Close < self.ll[hh_ll_Offset]
                            or self.window == 0.0
                    ):
                        self.position.close()
```

But for the purchase, it does not seem to operate within the specified 1 value: (the other operatorsm pm_v and self.pm_gap works and have a similar dynamic:

```

`            elif (self.window == 1.0) and (self.pm_gap < -pm_gap_val or self.pm_gap > pm_gap_val) and (pm_v_max > self.pm_v > pm_v_low):
                # Going Up:
                if crossover(self.data.High, self.pm_h) and (self.pd_h > self.pm_h):
                    if (self.pm_h + (atr_mult * self.atr)) > self.pd_h:
                        pass
                    elif self.buy(
                            stop=self.pm_h + (atr_mult * self.atr),
                            sl=price * (1 - 0.01),
                    ):
                        pass `
```
        
![Screenshot 2024-01-20 at 19 47 35](https://github.com/kernc/backtesting.py/assets/155924232/9dd9625d-ed4f-4ba5-be8e-fc31b9f4eb80)
